### PR TITLE
Add errorType to Categorise Debugger Errors

### LINKED ISSUE #6320: [Enhancement] Categorise Debugger Errors
## Description

Errors in the debugger today are of many kinds but unfortunately are all logged together with a common CTA that does not add much value.

## Types of Errors
- Validation error: These errors occur when there is a mismatch of data types in a field. These errors are appsmith controlled, few in nature, and should trigger the omnibar search.
- JS & Lint Errors: These errors are generated by the browser and not appsmith specific. These errors should lead users to Google to search for the error.
- Plugin execution Errors: These errors occur during action execution and can be due to the following reasons
  - Incorrect Datasource Configuration: This error is due to incorrect data source configuration and should trigger an intercom message with the error 
  - User action config Errors: This error is due to an incorrect configuration in the query and should search the omnibar for the solution.
  
Errors that do not fall into these buckets should trigger intercom

### PR DESCRIPTION
**Pull request by AnaghHegde**
## Description

> Plugin execution Errors: These errors occur during action execution and can be due to the following reasons
Incorrect Datasource Configuration: This error is due to incorrect data source configuration and should trigger an intercom message with the error
User action config Errors: This error is due to an incorrect configuration in the query and should search the omnibar for the solution

Related to  #6320 

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Local via postman
- Test B

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by AnaghHegde** [comment]
@akash-codemonk Let me know if this looks good. I will request for review after your confirmation.

**Review by sharat87 — COMMENTED** [review]
If I understand this correctly, we are passing on the error's enum name as is to the client right? So the client will look at the error name (called `errorType` for the client), and decide whether this error should open the Omnibar or a support conversation, correct?

If this is correct, can we do that logic in the backend instead? Let's only pass information to the client that'll help determine what to do with the error. Something like `errorType: "NEED_SUPPORT"` and `errorType: "SHOW_DOCS"` or similar.

The reason I'm inclined towards that is, one, it will avoid having the error names hard-coded in the client code. Otherwise, refactoring error names in the future will become fragile. Two, client has less intelligence to deal with, which is always desirable.

What say? Sorry if I'm misunderstanding the problem being solved here.

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/ActionExecutionResult.java [hunk 1]
```diff
@@ -38,9 +39,16 @@
...
+        if (error instanceof AppsmithPluginException) {
+            this.statusCode = ((AppsmithPluginException) error).getAppErrorCode().toString();
+            this.title = ((AppsmithPluginException) error).getTitle();
+            this.errorType = ((AppsmithPluginException) error).getErrorType();
+        }
+
+        else if (error instanceof BaseException) {
             this.statusCode = ((BaseException) error).getAppErrorCode().toString();
             this.title = ((BaseException) error).getTitle();
+            this.errorType = ((AppsmithPluginException) error).getErrorType();
```
Comment (sharat87) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/ActionExecutionResult.java:51]: Did you mean to use `BaseException` here?

```suggestion
            this.errorType = ((BaseException) error).getErrorType();
```

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/exceptions/pluginExceptions/AppsmithPluginException.java [hunk 2]
```diff
@@ -35,4 +35,6 @@ public AppsmithErrorAction getErrorAction() {
     }
 
     public String getTitle() { return this.error.getTitle(); }
+
+    public String getErrorType() { return this.error.name(); }
```
Comment (sharat87) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/external/exceptions/pluginExceptions/AppsmithPluginException.java:39]: The conversion of an `ErrorType` enum value to a `String` is being done by calling `.name()` here, and by calling `.toString()` in `AppsmithPluginError` enum. Is this difference intentional? If not, can we use the same method to stay consistent?

**Review by sharat87 — CHANGES_REQUESTED** [review]
Looks good. Just a couple of things before we can put this in.

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/ActionExecutionResult.java (hunk 1 above)
Comment (AnaghHegde) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/ActionExecutionResult.java:51]: @sharat87 `BaseException` class doesnt have the getErrorType(). Maybe its good to leave the errorType to empty when the exception is not of `Appsmith` type.

Comment (sharat87) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/ActionExecutionResult.java]: That may be but if the code flow has arrived at this point, that means `error instanceof AppsmithPluginException` was `false`, which means this typecast would definitely throw a `ClassCastException`. That was why I was concerned with this line. 😄

**Review by sharat87 — DISMISSED** [review]

**Review by sharat87 — APPROVED** [review]

**Comment by sharat87** [comment]
Looks good to be merged. Please squash merge once the CI is finished. Thanks!

**Comment by AnaghHegde** [comment]
@sharat87 Please approve again. Just fixed the merge conflicts :)

**Review by sharat87 — APPROVED** [review]

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/ErrorType.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package com.appsmith.external.models;
+
+import lombok.Getter;
+
+@Getter
+public enum ErrorType {
+    ARGUMENT_ERROR, CONFIGURATION_ERROR, CONNECTIVITY_ERROR, AUTHENTICATION_ERROR, API_ERROR, BAD_REQUEST, INTERNAL_ERROR
```
Comment (mohanarpit) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/ErrorType.java:7]: Super minor formatting nit: Please define enum values in new lines. Makes it readable & parseable to the naked eye.

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/ActionExecutionResult.java [hunk 4]
```diff
@@ -38,7 +39,13 @@
...
         this.body = error.getMessage();
 
-        if (error instanceof BaseException) {
+        if (error instanceof AppsmithPluginException) {
+            this.statusCode = ((AppsmithPluginException) error).getAppErrorCode().toString();
+            this.title = ((AppsmithPluginException) error).getTitle();
+            this.errorType = ((AppsmithPluginException) error).getErrorType();
+        }
+
+        else if (error instanceof BaseException) {
```
Comment (mohanarpit) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/ActionExecutionResult.java:48]: Super minor nit: Always have the else clause at the end of the if-block. This ensures that as a reader, I notice the else immediately. If the else block is removed from the if block, it makes code slightly harder to read.
```suggestion
        } else if (error instanceof BaseException) {
```

**Comment by AnaghHegde** [comment]
@mohanarpit ready for review

**Review by mohanarpit — APPROVED** [review]

**Pull request merged**